src/ChluIPFS.js

```diff
@@ -0,0 +1,184 @@
+        }
+        // OrbitDB setup
+        if (!this.orbitDb) {
+            this.orbitDb = new OrbitDB(this.ipfs, this.orbitDbDirectory);
+        }
+        if (this.type === constants.types.customer && !this.db) {
+            this.db = await this.orbitDb.feed('chlu-experimental-customer-review-updates');
+        }
+        // PubSub setup
+        if (!this.room) {
+            this.room = Room(this.ipfs, 'chlu-experimental');
+            // handle events
+            this.room.on('message', message => this.handleMessage(message));
+            // wait for room subscription
+            await new Promise(resolve => {
+                this.room.on('subscribed', () => resolve());
+            });
+        }
+        // If customer, also wait for at least one peer to join the room (TODO: review this)
+        if (this.type === constants.types.customer) {
+            if (this.room.getPeers().length === 0) {
+                await new Promise(resolve => {
+                    this.room.on('peer joined', () => resolve());
+                });
+            }
+        } else if (this.type === constants.types.service) {
+            this.startServiceNode();
+        }
+        return true;
+    }
+
+    async stop() {
+        // leaving the pubsub room is already handled by the library on IPFS stop
+        if (this.db) await this.db.close();
+        await this.ipfs.stop();
+        this.db = undefined;
+        this.room = undefined;
+        this.orbitDb = undefined;
+        this.ipfs = undefined;
+    }
+
+    async pin(multihash){
+        // broadcast start of pin process
+        await this.room.broadcast(this.utils.encodeMessage({ type: constants.eventTypes.pinning, multihash }));
+        if (this.ipfs.pin) {
+            await this.ipfs.pin.add(multihash, { recursive: true });
+        } else {
+            // TODO: Chlu service node need to be able to pin, so we should support using go-ipfs
+            logger.warn('This node is running an IPFS client that does not implement pinning. Falling back to just retrieving the data non recursively. This will not be supported');
+            await this.ipfs.object.get(multihash);
+        }
+        // broadcast successful pin
+        await this.room.broadcast(this.utils.encodeMessage({ type: constants.eventTypes.pinned, multihash }));
+    }
+
+    getOrbitDBAddress(){
+        if (this.type === constants.types.customer) {
+            return this.db.address.toString();
+        } else {
+            throw new Error('Not a customer');
+        }
+    }
+
+    async storeReviewRecord(reviewRecord){
+        // TODO: check format, check is customer
+        // write thing to ipfs
+        const dagNode = await this.ipfs.object.put(reviewRecord);
+        const multihash = this.utils.multihashToString(dagNode.multihash);
+        // Broadcast request for pin, then wait for response
+        // TODO: handle a timeout and also rebroadcast periodically, otherwise new peers won't see the message
```

Comment:
It would be good to have a timeout as you said, but maybe use an exponential backoff while retrying. I am surprised orbitdb doesn't do that already

Reply:
I never thought about retrying but it makes a lot of sense

Reply:
I opened an issue about this: https://github.com/ChluNetwork/chlu-ipfs-support/issues/6